src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+     * @returns {number} Result of comparison
+     */
+    function _compareMatchResultToPos(matchIndex, posIndex) {
+        if (matchIndex === posIndex) {
+            return 0;
+        } else if (matchIndex < posIndex) {
+            return -1;
+        } else {
+            return 1;
+        }
+    }
+
+    /**
+     * Finds the result that is at or nearest the position passed to function.
+     * If a match result is not at the position, it will then locate the closest
+     * match result which is in the search direction.
+     * If there is no match found before the end or beginning of the document
+     * then this function returns false.
+     * @param {!Object} regexIndexer instance of the regex indexer. see _createRegexIndexer
+     * @param {number} pos Starting index position to search from
+     * @param {boolean} reverse direction to search.
+     * @param {function(number, number)} fnCompare function to compare positions for binary search
+     */
+    function _findResultIndexNearPos(regexIndexer, pos, reverse, fnCompare) {
+        var compare;
+
+        var length = regexIndexer.getItemCount();
+        var upperBound = length - 1;
+        var lowerBound = 0;
+        var searchIndex;
+        while (lowerBound <= upperBound) {
+            searchIndex = Math.floor((upperBound + lowerBound) / 2);
+            compare = fnCompare(regexIndexer.getMatchIndexStart(searchIndex), pos);
+            if (compare === 0) {
+                return searchIndex;
+            } else if (compare === -1) {
+                lowerBound = searchIndex + 1;
+            } else {
+                upperBound = searchIndex - 1;
+            }
+        }
+        // no exact match, we are at the lower bound
+        // if going forward return the next index
+        if ((compare === -1) && (!reverse)) {
+            searchIndex += 1;
+        }
+        // no exact match, we are at the upper bound
+        // if going reverse return the next lower index
+        if ((compare === 1) && (reverse)) {
+            searchIndex -= 1;
+        }
+
+        // If we went beyond the length or start, there was no match and no next index to match
+        if ((searchIndex < 0) || (searchIndex >= length)) {
+            return false;
+        }
+        // no exact match, we are already at the closest match in the search direction
+        return searchIndex;
+    }
+
+    /**
+     * Enhances array with functions which facilitate managing the array contents
+     * by groups of items.
+     * This is useful for both performance and memory consumption to store the indexes
+     * of the match result beginning and ending locations.
+     * @param {Array} array The array to enhance
+     * @param {number} groupSize The number of indices that belong to a group
+     * @returns {GroupArray} Enhanced Array
+     */
+    function _makeGroupArray(array, groupSize) {
+        var _currentGroupIndex = -groupSize;
+        _.assign(array, {
```

Comment:
May be it could be a better implementation if `GroupArray` was a real datatype (function) which would contain its internal state. The same is true for the "indexer" and the cursor objects.

Instead of using `_make...` functions and `_.assign` or `_.extend`. You'll have a constructor and use `new`.

For example, for `GroupArray` you'd use:

```javascript
function GroupArray(array, groupSize) {
    this._currentGroupIndex = -groupSize;
    this._size = groupSize;
    this._array = array;
}

GroupArray.prototype = Object.create(null); // do not inherit from object
GroupArray.prorotype.constructor = GroupArray;

GroupArray.prototype.nextGroupIndex = function () {
    // your implementation referring to this instead of make function context...
};
```